include/dg/llvm/ValueRelations/RelationsAnalyzer.hpp

```diff
@@ -0,0 +1,1093 @@
+#include <llvm/IR/IntrinsicInst.h>
+#include <llvm/IR/Constants.h>
+#include <llvm/IR/Module.h>
+#include <llvm/IR/CFG.h>
+
+#if (__clang__)
+#pragma clang diagnostic pop // ignore -Wunused-parameter
+#else
+#pragma GCC diagnostic pop
+#endif
+
+#include <algorithm>
+#include <string>
+
+#include "GraphElements.hpp"
+#include "ValueRelations.hpp"
+#include "StructureAnalyzer.hpp"
+
+#ifndef NDEBUG
+#include "getValName.h"
+#endif
+
+namespace dg {
+namespace vr {
+
+
+class RelationsAnalyzer {
+
+    const std::set<std::string> safeFunctions = { "__VERIFIER_nondet_int", "__VERIFIER_nondet_char" };
+
+    const llvm::Module& module;
+
+    // VRLocation corresponding to the state of the program BEFORE executing the instruction
+    const std::map<const llvm::Instruction *, VRLocation *>& locationMapping;
+    const std::map<const llvm::BasicBlock *, std::unique_ptr<VRBBlock>>& blockMapping;
+
+    // holds information about structural properties of analyzed module
+    // like set of instructions executed in loop starging at given location
+    // or possibly set of values defined at given location
+    const StructureAnalyzer& structure;
+
+    bool processOperation(VRLocation* source, VRLocation* target, VROp* op) {
+        if (! target) return false;
+        assert(source && target && op);
+
+        ValueRelations newGraph = source->relations;
+
+        if (op->isInstruction()) {
+            const llvm::Instruction* inst = static_cast<VRInstruction *>(op)->getInstruction();
+            forgetInvalidated(newGraph, inst);
+            processInstruction(newGraph, inst);
+        } else if (op->isAssume()) { 
+            if (op->isAssumeBool())
+                processAssumeBool(newGraph, static_cast<VRAssumeBool *>(op));
+            else // isAssumeEqual
+                processAssumeEqual(newGraph, static_cast<VRAssumeEqual *>(op));
+        } // else op is noop
+
+        return andSwapIfChanged(target->relations, newGraph);
+    }
+
+    void forgetInvalidated(ValueRelations& graph, const llvm::Instruction* inst) const {
+
+        for (const llvm::Value* invalid : instructionInvalidatesFromGraph(graph, inst))
+            graph.unsetAllLoadsByPtr(invalid);
+    }
+
+    void addAndUnwrapLoads(
+            std::set<std::pair<const llvm::Value*, unsigned>>& writtenTo,
+            const llvm::Value* val) const {
+
+        unsigned depth = 0;
+        writtenTo.emplace(val, 0);
+        while (auto load = llvm::dyn_cast<llvm::LoadInst>(val)) {
+            writtenTo.emplace(load->getPointerOperand(), ++depth);
+            val = load->getPointerOperand();
+        }
+    }
+
+    std::set<std::pair<const llvm::Value*, unsigned>> instructionInvalidates(const llvm::Instruction* inst) const {
+
+        if (! inst->mayWriteToMemory() && ! inst->mayHaveSideEffects())
+            return std::set<std::pair<const llvm::Value*, unsigned>>();
+
+        if (auto intrinsic = llvm::dyn_cast<llvm::IntrinsicInst>(inst)) {
```

Comment:
The switch could be factored out into a method that returns true on the right type of intrinsic  (not necessary, just a note)

Reply:
Ou, that is a good idea. Why did not I see that..? Anyway, done.